### PR TITLE
IfcArcIndex implementetion 

### LINKED ISSUE #68: Example file "beam-extruded-solid" not rendered.
The file is [here](../blob/development/testdata/IFC4x1ExampleFiles/official/E8_Building_element_standard_case/beam-extruded-solid.ifc). OIP doesn't show anything.

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Fixes #68

Trying to implement it based on https://math.stackexchange.com/questions/1379217/3-points-in-3d-space-to-find-the-center-of-an-arc-or-circle

And http://algolist.ru/maths/geom/equation/Circle.cpp

### CODE REVIEW

**Comment by SamuilsRulovs** [comment]
convertIfcArcIndex is working, however there are some mistakes in calculations, so the slab-standard-case is not correctly rendered yet.  
Screenshot 2020-11-22 193601

**Comment by SamuilsRulovs** [comment]
image

Now convertIfcIndexedPolyCurve works correctly

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 1]
```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
...
+						// TYPE IfcLineIndex = LIST [2:?] OF IfcPositiveInteger;
+						return convertIfcLineIndex(segmentIndexSelect.get<1>(), points);
+					}
+					default:
+						throw oip::UnhandledException();
+					}
+				}
+				
+				std::vector<carve::geom::vector<3>> convertIfcLineIndex(
+					const typename IfcEntityTypesT::IfcLineIndex &arcSegment,
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: Copy-paste mistake - arcIndex vs convertIfcLineIndex?

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 2]
```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
...
+					const std::vector<carve::geom::vector<3>> points) const throw(...)
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: please pass it as reference (`&`).

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 3]
```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
...
+						loop.push_back(points[i - 1]); //EXPRESS count from 1, C++ from 0
+					}
+					return loop;
+				}
+
+
+				
+				std::vector<carve::geom::vector<3>> convertIfcArcIndex(
+					const typename IfcEntityTypesT::IfcArcIndex &arcSegment, 
+					const std::vector<carve::geom::vector<3>> points) const throw(...)
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: reference

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 4]
```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
...
+					normalVector.normalize();
+					//u=(C−A)×n
+					carve::geom::vector<3> firstOrthogonalDirection = carve::geom::cross(arcEnd - arcStart, normalVector);
+					firstOrthogonalDirection.normalize();
+					//v=n×u
+					carve::geom::vector<3> secondOrthogonalDirection = carve::geom::cross(normalVector, firstOrthogonalDirection);
+					secondOrthogonalDirection.normalize();
+		
+					//Calculate distance of the plane to the origin
+					double distance = (normalVector.x*arcStart.x) + (normalVector.y*arcStart.y) + (normalVector.y*arcStart.y);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: Is this correct? why is `.x` plus `.y` plus `.y`? Why twice the `y`?

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 5]
```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
...
+
+					//Convert the problem into a 2D problem
+					carve::math::Matrix rotationMatrix = carve::math::Matrix(
+						firstOrthogonalDirection.x, secondOrthogonalDirection.x, 0, 0,
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]:
```suggestion
						firstOrthogonalDirection.x, secondOrthogonalDirection.x, 0., 0.,
```
And similar to all entries to the matrix.

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 6]
```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
...
+							loop_intern.push_back(carve::geom::VECTOR(
+								normalVector.x * distance + firstOrthogonalDirection.x * circle_points[i].x + secondOrthogonalDirection.x * circle_points[i].y,
+								normalVector.y * distance + firstOrthogonalDirection.y * circle_points[i].x + secondOrthogonalDirection.y * circle_points[i].y,
+								normalVector.z * distance + firstOrthogonalDirection.z * circle_points[i].x + secondOrthogonalDirection.z * circle_points[i].y));
+						}
+						return loop_intern;
+					}
+				}
+
+				carve::geom::vector<2> covert3Dto2D(carve::math::Matrix& conversionMatrix, carve::geom::vector<3>& vector3D) const throw(...){
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]:
```suggestion
				carve::geom::vector<2> convert3Dto2D(const carve::math::Matrix& conversionMatrix, const carve::geom::vector<3>& vector3D) const throw(...){
```

**Review by pjanck — CHANGES_REQUESTED** [review]
Small changes requested.

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 7]
```diff
@@ -833,14 +863,14 @@ namespace OpenInfraPlatform {
...
 
 					//Convert the problem into a 2D problem
 					carve::math::Matrix rotationMatrix = carve::math::Matrix(
-						firstOrthogonalDirection.x, secondOrthogonalDirection.x, 0, 0,
-						firstOrthogonalDirection.y, secondOrthogonalDirection.y, 0, 0,
-						firstOrthogonalDirection.z, secondOrthogonalDirection.z, 1, 0,
-						0, 0, 0, 1);
+						firstOrthogonalDirection.x, secondOrthogonalDirection.x, 0., 0.,
+						firstOrthogonalDirection.y, secondOrthogonalDirection.y, 0., 0.,
+						firstOrthogonalDirection.z, secondOrthogonalDirection.z, 1, 0.,
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]:
```suggestion
						firstOrthogonalDirection.z, secondOrthogonalDirection.z, 1., 0.,
```

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 8]
```diff
@@ -833,14 +863,14 @@ namespace OpenInfraPlatform {
 					secondOrthogonalDirection.normalize();
 		
 					//Calculate distance of the plane to the origin
-					double distance = (normalVector.x*arcStart.x) + (normalVector.y*arcStart.y) + (normalVector.y*arcStart.y);
+					double distance = (normalVector.x*arcStart.x) + (normalVector.y*arcStart.y) + (normalVector.z*arcStart.z);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: This hsould probably also work as `normalVector * arcStart`

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 9]
```diff
@@ -900,7 +930,13 @@ namespace OpenInfraPlatform {
...
 				}
 
-				carve::geom::vector<2> covert3Dto2D(carve::math::Matrix& conversionMatrix, carve::geom::vector<3>& vector3D) const throw(...){
+				/**********************************************************************************************/
+				/*! \brief Calculates coordinates of the intersection point.
+				* \param[in] conversionMatrix		A pointer to data from IfcIndexedPolyCurve.
+				* \param[in] vector3D				Vector in 3D
+				* return							converterted vector in 2D. 		
+				*/
+				carve::geom::vector<2> covert3Dto2D(const carve::math::Matrix&  conversionMatrix, const carve::geom::vector<3>& vector3D) const throw(...){
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]:
```suggestion
				carve::geom::vector<2> convert3Dto2D(const carve::math::Matrix&  conversionMatrix, const carve::geom::vector<3>& vector3D) const throw(...){
```

File: UnitTests/Schemas/IFC4X1/slab-standard-case/src/slab-standard-case.cpp [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcImporterT<emt::IFC4X1EntityTypes>> importer = nullptr;
+	buw::ReferenceCounted<oip::IfcModel> model = buw::makeReferenceCounted<oip::IfcModel>();
+};
+
+TEST_F(SlabStandardCase, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(38));
+}
+
+TEST_F(SlabStandardCase, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCINDEXEDPOLYCURVE"; });
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/slab-standard-case/src/slab-standard-case.cpp]: Additionally, check for IFCEXTRUDEDAREASOLID and IFCARBITRARYCLOSEDPROFILEDEF

**Review by pjanck — CHANGES_REQUESTED** [review]
Doxygen comments need to be adjusted as well.

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 6 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: f9547d9

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 5 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: f9547d9

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 4 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: f9547d9

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 3 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: f9547d9

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 2 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: f9547d9

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 1 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: f9547d9

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 8 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Unfortunately, this doesn’t work

File: UnitTests/Schemas/IFC4X1/slab-standard-case/src/slab-standard-case.cpp (hunk 10 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/slab-standard-case/src/slab-standard-case.cpp]: Resolved: bb58436

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 9 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: bb58436

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 8 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: bb58436

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 7 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: bb58436

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 11]
```diff
@@ -936,7 +936,7 @@ namespace OpenInfraPlatform {
 				* \param[in] vector3D				Vector in 3D
 				* return							converterted vector in 2D.
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: Typo?

**Review by pjanck — CHANGES_REQUESTED** [review]
Small typo in Doxycomment, otherwise LFTM.

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 11 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: 4c2f2f4

**Review by pjanck — APPROVED** [review]

**Pull request merged**